src/linker/Linker/TypeReferenceMarker.cs

```diff
@@ -0,0 +1,343 @@
+
+namespace Mono.Linker
+{
+	struct TypeReferenceMarker
+	{
+		readonly AssemblyDefinition assembly;
+		readonly MarkingHelpers markingHelpers;
+		HashSet<TypeReference> visited;
+
+		public TypeReferenceMarker (AssemblyDefinition assembly, MarkingHelpers markingHelpers)
+		{
+			this.assembly = assembly;
+			this.markingHelpers = markingHelpers;
+			visited = null;
+		}
+
+		// Traverse the assembly and mark the scopes of discovered type references (but not exported types).
+		// This includes scopes referenced by Cecil TypeReference objects that don't represent rows in the typeref table,
+		// such as references to built-in types, or attribute arguments which encode type references as strings.
+		public void Process ()
+		{
+			visited = new HashSet<TypeReference> ();
+
+			WalkCustomAttributesTypesScopes (assembly);
+			WalkSecurityAttributesTypesScopes (assembly);
+
+			foreach (var module in assembly.Modules)
+				WalkCustomAttributesTypesScopes (module);
+
+			var mmodule = assembly.MainModule;
+			if (mmodule.HasTypes) {
+				foreach (var type in mmodule.Types) {
+					WalkScopes (type);
+				}
+			}
+
+			visited = null;
+		}
+
+		void WalkScopes (TypeDefinition typeDefinition)
+		{
+			WalkCustomAttributesTypesScopes (typeDefinition);
+			WalkSecurityAttributesTypesScopes (typeDefinition);
+
+			if (typeDefinition.BaseType != null)
+				WalkScopeOfTypeReference (typeDefinition.BaseType);
+
+			if (typeDefinition.HasInterfaces) {
+				foreach (var iface in typeDefinition.Interfaces) {
+					WalkCustomAttributesTypesScopes (iface);
+					WalkScopeOfTypeReference (iface.InterfaceType);
+				}
+			}
+
+			if (typeDefinition.HasGenericParameters)
+				WalkTypeScope (typeDefinition.GenericParameters);
+
+			if (typeDefinition.HasEvents) {
+				foreach (var e in typeDefinition.Events) {
+					WalkCustomAttributesTypesScopes (e);
+					// e.EventType is not saved
+				}
+			}
+
+			if (typeDefinition.HasFields) {
+				foreach (var f in typeDefinition.Fields) {
+					WalkCustomAttributesTypesScopes (f);
+					WalkScopeOfTypeReference (f.FieldType);
+					WalkMarshalInfoTypeScope (f);
+				}
+			}
+
+			if (typeDefinition.HasMethods) {
+				foreach (var m in typeDefinition.Methods) {
+					WalkCustomAttributesTypesScopes (m);
+					WalkSecurityAttributesTypesScopes (m);
+					if (m.HasGenericParameters)
+						WalkTypeScope (m.GenericParameters);
+
+					WalkCustomAttributesTypesScopes (m.MethodReturnType);
+					WalkScopeOfTypeReference (m.MethodReturnType.ReturnType);
+					WalkMarshalInfoTypeScope (m.MethodReturnType);
+					if (m.HasOverrides) {
+						foreach (var mo in m.Overrides)
+							WalkMethodReference (mo);
+					}
+
+					if (m.HasParameters)
+						WalkTypeScope (m.Parameters);
+
+					if (m.HasBody)
+						WalkTypeScope (m.Body);
+				}
+			}
+
+			if (typeDefinition.HasProperties) {
+				foreach (var p in typeDefinition.Properties) {
+					WalkCustomAttributesTypesScopes (p);
+					// p.PropertyType is not saved
```

Comment:
For future reference - this feels dangerous - not because of saving, but because of custom steps. We update the type refs in sweep step - but we don't update this one (intentionally). Again - nothing to do with this change, just that I noticed this...